### DISCUSSION TITLE
Contact & ASN migration scripts for NetBox v3.2

### DISCUSSION THREAD

**Discussion by jeremystretch** [netbox-community/netbox#8975] [General]
Hey folks! As you may know, NetBox v3.2 is removing the legacy contact and ASN fields from the site model, in favor of the new independent Contact and ASN objects introduced in v3.1. A few people have asked about scripts to help migrate legacy data, and I finally got some time today to work on this.

I've published a module with two custom scripts, available here: https://github.com/netbox-community/migration-scripts These scripts can be installed in your current NetBox v3.1 deployment to assist in automatically migrating data from the legacy site fields to the new objects.

![migrate_contacts](https://user-images.githubusercontent.com/13487278/160190435-6351bca0-9bac-46d3-810f-6cccb34577c9.png)

![migrate_contacts_log](https://user-images.githubusercontent.com/13487278/160191124-73208524-9d87-4511-b7fd-428ea9dbba2f.png)

---

![migrate_asns](https://user-images.githubusercontent.com/13487278/160190439-2c690fbd-5ccb-4f38-a545-e7862a7ca62e.png)

![migrate_asns_log](https://user-images.githubusercontent.com/13487278/160191233-4abd4124-93d9-4f8a-9499-01ffb12dd32c.png)

The scripts should address simple use cases fairly well, but as always tread with caution and do plenty of dry-runs.

Users are also strongly encouraged to modify local copies of these scripts to better suit their own needs. For example, you might want to exclude sites with certain roles, or skip a range of AS numbers that aren't used anymore. (Be sure to check out NetBox's [custom scripts documentation](https://netbox.readthedocs.io/en/stable/customization/custom-scripts/) before getting started.)

Anyway, hopefully these will come in handy as people get ready for NetBox v3.2!

**Comment by TrixaFett** [comment]
Both scritpts fail.
Here fail of the Migrate Contacts:

<html>
<body>


Failure | An exception occurred: FieldError: Cannot resolve keyword 'contact_name' into field. Choices are: _name, asns, circuit_terminations, clusters, comments, contacts, created, custom_field_data, description, devices, facility, group, group_id, id, images, journal_entries, last_updated, latitude, locations, longitude, name, physical_address, powerpanel, prefixes, racks, region, region_id, shipping_address, slug, status, tagged_items, tags, tenant, tenant_id, time_zone, vlan_groups, vlansTraceback (most recent call last):   File "/opt/netbox/netbox/extras/scripts.py", line 461, in _run_script     script.output = script.run(data=data, commit=commit)   File "/etc/netbox/scripts/netbox_v32_migration.py", line 61, in run     sites = Site.objects.exclude(contact_name='')   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/manager.py", line 85, in manager_method     return getattr(self.get_queryset(), name)(*args, **kwargs)   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/query.py", line 1079, in exclude     return self._filter_or_exclude(True, args, kwargs)   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/query.py", line 1089, in _filter_or_exclude     clone._filter_or_exclude_inplace(negate, args, kwargs)   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/query.py", line 1094, in _filter_or_exclude_inplace     self._query.add_q(~Q(*args, **kwargs))   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/sql/query.py", line 1502, in add_q     clause, _ = self._add_q(q_object, self.used_aliases)   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/sql/query.py", line 1532, in _add_q     child_clause, needed_inner = self.build_filter(   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/sql/query.py", line 1377, in build_filter     lookups, parts, reffed_expression = self.solve_lookup_type(arg)   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/sql/query.py", line 1187, in solve_lookup_type     _, field, _, lookup_parts = self.names_to_path(lookup_splitted, self.get_meta())   File "/opt/netbox/venv/lib/python3.10/site-packages/django/db/models/sql/query.py", line 1677, in names_to_path     raise FieldError( django.core.exceptions.FieldError: Cannot resolve keyword 'contact_name' into field. Choices are: _name, asns, circuit_terminations, clusters, comments, contacts, created, custom_field_data, description, devices, facility, group, group_id, id, images, journal_entries, last_updated, latitude, locations, longitude, name, physical_address, powerpanel, prefixes, racks, region, region_id, shipping_address, slug, status, tagged_items, tags, tenant, tenant_id, time_zone, vlan_groups, vlans
-- | --


<br class="Apple-interchange-newline">
</body>
</html>

**Reply by jeremystretch** [reply]
You need to run them on NetBox v3.1, _before_ upgrading to v3.2.

**Reply by TrixaFett** [reply]
We are running 3.1.11 and that is when it fails.

**Reply by DanSheps** [reply]
> We are running 3.1.11 and that is when it fails.

This looks to not be the case.  Both contact_name and other fields are still in for 3.1.11.

**Reply by TrixaFett** [reply]
I found the mistake and it was my fault, when we looked behind the scenes there were mismatches in the database version.

**Comment by stoms12** [comment]
We are having an issue with running these scripts too:
Running NetBox v3.1.11-ldap-1.6.1 in docker
``` 
(venv) bash-5.1$ python3 manage.py runscript --loglevel debug --data '{"asn_rir": 2, "clear_site_field": false}' netbox_v32_migration.MigrateSiteASNsScript
🧬 loaded config '/etc/netbox/config/configuration.py'
🧬 loaded config '/etc/netbox/config/extra.py'
🧬 loaded config '/etc/netbox/config/logging.py'
🧬 loaded config '/etc/netbox/config/plugins.py'

[2022-08-25 09:27:08,332][INFO] - Running script (commit=False)
[2022-08-25 09:27:08,334][INFO] - Found 2 sites with a legacy ASN defined.
[2022-08-25 09:27:08,336][INFO] - Creating new ASN: 16509
[2022-08-25 09:27:08,337][ERROR] - An exception occurred: `ValueError: Cannot assign "2": "ASN.rir" must be a "RIR" instance.`
---
Traceback (most recent call last):
  File "/opt/netbox/netbox/extras/management/commands/runscript.py", line 44, in _run_script
    script.output = script.run(data=data, commit=commit)
  File "/etc/netbox/scripts/netbox_v32_migration.py", line 150, in run
    asn = ASN(asn=site.asn, rir=data['asn_rir'])
  File "/opt/netbox/venv/lib/python3.9/site-packages/django/db/models/base.py", line 485, in __init__
    _setattr(self, field.name, rel_obj)
  File "/opt/netbox/venv/lib/python3.9/site-packages/django/db/models/fields/related_descriptors.py", line 215, in __set__
    raise ValueError(
ValueError: Cannot assign "2": "ASN.rir" must be a "RIR" instance.

---
[2022-08-25 09:27:08,337][INFO] - Database changes have been reverted due to error.
[2022-08-25 09:27:08,337][ERROR] - Exception raised during script execution: Cannot assign "2": "ASN.rir" must be a "RIR" instance.
[2022-08-25 09:27:08,340][INFO] - Script completed in 0 minutes, 0.01 seconds
```

Also seeing the same issue with the contact migration script

```
(venv) bash-5.1$ python3 manage.py runscript --loglevel debug --data '{"contact_role": 1, "contact_priority": "primary", "clear_site_fields": false}' netbox_v32_migration.MigrateSiteContactsScript
🧬 loaded config '/etc/netbox/config/configuration.py'
🧬 loaded config '/etc/netbox/config/extra.py'
🧬 loaded config '/etc/netbox/config/logging.py'
🧬 loaded config '/etc/netbox/config/plugins.py'
[2022-08-25 09:31:59,190][INFO] - Running script (commit=False)
[2022-08-25 09:31:59,193][INFO] - Found 9 sites with legacy contact information defined.
[2022-08-25 09:31:59,195][INFO] - Creating new contact: John Smith
[2022-08-25 09:31:59,292][INFO] - Assigning contact John Smith to site Johns House
[2022-08-25 09:31:59,294][ERROR] - An exception occurred: `ValueError: Cannot assign "1": "ContactAssignment.role" must be a "ContactRole" instance.`
---
Traceback (most recent call last):
  File "/opt/netbox/netbox/extras/management/commands/runscript.py", line 44, in _run_script
    script.output = script.run(data=data, commit=commit)
  File "/etc/netbox/scripts/netbox_v32_migration.py", line 87, in run
    assignment = ContactAssignment(
  File "/opt/netbox/venv/lib/python3.9/site-packages/django/db/models/base.py", line 485, in __init__
    _setattr(self, field.name, rel_obj)
  File "/opt/netbox/venv/lib/python3.9/site-packages/django/db/models/fields/related_descriptors.py", line 215, in __set__
    raise ValueError(
ValueError: Cannot assign "1": "ContactAssignment.role" must be a "ContactRole" instance.

---
[2022-08-25 09:31:59,294][INFO] - Database changes have been reverted due to error.
[2022-08-25 09:31:59,294][ERROR] - Exception raised during script execution: Cannot assign "1": "ContactAssignment.role" must be a "ContactRole" instance.
[2022-08-25 09:31:59,298][INFO] - Script completed in 0 minutes, 0.11 seconds
```
